src/include/machnet_engine.h

```diff
@@ -874,7 +962,73 @@ class MachnetEngine {
+            if (qid > -1) {
+              auto *response = CHECK_NOTNULL(packet_pool_->PacketAlloc());
+              const size_t hdr_length =
+                  (sizeof(Ethernet) + sizeof(Ipv4) + sizeof(Udp) +
+                   sizeof(net::MachnetPktHdr));
+              const uint32_t pkt_len =
+                  hdr_length +
+                  46;  // I am assuming the rss key size to 40 bytes.
+              Ethernet *response_eh = response->append<Ethernet *>(pkt_len);
+
+              // THIS IS A BIG PROBLEM!
+              LOG(WARNING) << pkt->length() << " XXXX " << pkt_len;
+              if (!response_eh) LOG(WARNING) << "could not append";
+              // Size is different that just packet->length;
+              response_eh->dst_addr = eh->src_addr;
+              response_eh->src_addr = pmd_port_->GetL2Addr();
+              response_eh->eth_type = be16_t(Ethernet::kIpv4);
+              response->set_l2_len(sizeof(*response_eh));
+              Ipv4 *response_ipv4h = reinterpret_cast<Ipv4 *>(response_eh + 1);
+              response_ipv4h->version_ihl = 0x45;
+              response_ipv4h->type_of_service = 0;
+              response_ipv4h->packet_id = be16_t(0x1513);
+              response_ipv4h->fragment_offset = be16_t(0);
+              response_ipv4h->next_proto_id = Ipv4::Proto::kUdp;
+              response_ipv4h->time_to_live = 64;
+              // response_ipv4h->total_length = be16_t(pkt->length() -
+              // sizeof(Ethernet));
+              response_ipv4h->total_length = be16_t(pkt_len - sizeof(Ethernet));
+              response_ipv4h->src_addr = ipv4h->dst_addr;
+              response_ipv4h->dst_addr = ipv4h->src_addr;
+              response_ipv4h->hdr_checksum = 0;
+              response->set_l3_len(sizeof(*response_ipv4h));
+              response->offload_ipv4_csum();  // I should calculate the checksum
+              Udp *response_udp = reinterpret_cast<Udp *>(response_ipv4h + 1);
+              response_udp->dst_port = udph->src_port;
+              response_udp->src_port = udph->dst_port;
+              // response_udp->len = be16_t(pkt->length() - sizeof(Ethernet) -
+              // sizeof(Ipv4));
+              response_udp->len =
+                  be16_t(pkt_len - sizeof(Ethernet) - sizeof(Ipv4));
+              response_udp->cksum = be16_t(0);
+              // calculate checksum for udp;
+              net::MachnetPktHdr *response_mach =
+                  reinterpret_cast<net::MachnetPktHdr *>(response_udp + 1);
+              response_mach->magic = be16_t(net::MachnetPktHdr::kMagic);
+              response_mach->net_flags =
+                  net::MachnetPktHdr::MachnetFlags::kRetryForRss;
+              response_mach->msg_flags = 0;
+              response_mach->seqno = be32_t(0);
+              response_mach->ackno = be32_t(0);
+              response_mach->sack_bitmap = be64_t(0);
+              response_mach->sack_bitmap_count = be16_t(0);
```

Comment:
This is incorrect handling. If the client tries to open two connections, one for an application that listens to the local engine, and one other that is for a sibling engine how could the remote client match your response if there are no sequence number matching?

Reply:
You are correct, the src port is going to save it in this example.

Still I think we should treat `kRetry` as a control packet, similar to `kRst`. Checking sequence numbers should be safer and cleaner, and wouldn't require many changes!